src/type_methods.jl

```diff
@@ -0,0 +1,106 @@
+
+# build expression for tuples of the parameterizations data
+function expr_gen(typed, fields)
+    expr = :(())  # build an expression for a tuple
+    acc = typed ? typed_acc : untyped_acc
+    append!(expr.args, [acc(field) for field in fields])
+    return expr
+end
+untyped_acc(i) = isa(i, Int) ? :(X[$(i)]) : :(X.$(i))
+typed_acc(i) = isa(i, Int) ? :(T(X[$(i)])) : :(T(X.$(i)))
+
+
+#
+# function to create a code block to convert to and from Vec and Vector rtypes
+#
+function add_methods(rot_type)
+
+    #
+    # Build expressions for building typed and untyped tuples using an Integer Index
+    #
+    untyped_indexed_expr = expr_gen(false, 1:numel(rot_type))
+    typed_indexed_expr = expr_gen(true, 1:numel(rot_type))
+
+    #
+    # Build expressions for building typed and untyped tuples using a field
+    #
+
+    fields = fieldnames(rot_type)
+    untyped_field_expr = (rot_type <: Mat) ? untyped_indexed_expr : expr_gen(false, fields[1:numel(rot_type)])
+    typed_field_expr = (rot_type <: Mat) ? typed_indexed_expr : expr_gen(true, fields[1:numel(rot_type)])
+
+
+    # the code block to return
+    qb = quote end
+
+    #
+    # convert to / from a mutable vector
+    #
+    qn = quote
+
+        # convert to a mutable vector
+        @inline convert(::Type{Vector}, X::$(rot_type)) = vcat($(untyped_field_expr.args...))
+        @inline convert{T <: Real}(::Type{Vector{T}}, X::$(rot_type)) = vcat($(typed_field_expr.args...))
+        @inline call{T <: Vector}(::Type{T}, X::$(rot_type)) = convert(T, X)
+
+        # convert from a mutable vector
+        @inline convert{T <: $(rot_type), U <: Real}(::Type{T}, X::Vector{U}) = T($(untyped_indexed_expr.args...))
+        @inline call{T <: $(rot_type), U <: Real}(::Type{T}, X::Vector{U}) = convert(T, X)
+
+        # convert to an immutable vector
+        @inline convert(::Type{Vec}, X::$(rot_type)) = Vec($(untyped_field_expr.args...))
+        @inline convert{T <: Real}(::Type{Vec{$(numel(rot_type)), T}}, X::$(rot_type)) = Vec($(typed_field_expr.args...))
+        @inline call(::Type{Vec}, X::$(rot_type)) = convert(Vec, X)
+        @inline call{T <: Real}(::Type{Vec{$(numel(rot_type)), T}}, X::$(rot_type)) = convert(Vec{$(numel(rot_type)), T}, X)
+
+        # convert from an immutable vector
+        @inline convert{T <: $(rot_type), U <: Real}(::Type{T}, X::FixedVector{$(numel(rot_type)), U}) = T($(untyped_indexed_expr.args...))
+        @inline call{T <: $(rot_type), U <: Real}(::Type{T}, X::FixedVector{$(numel(rot_type)), U}) = convert(T, X)
+
+        # convert to a tuple
+        # @inline tuple(X::$(rot_type)) = tuple($(untyped_field_expr.args...))
+
+        # convert from a tuple
+        @inline convert{T <: $(rot_type), U <: Real}(::Type{T}, X::NTuple{$(numel(rot_type)), U}) = T($(untyped_indexed_expr.args...))
+        @inline call{T <: $(rot_type), U <: Real}(::Type{T}, X::NTuple{$(numel(rot_type)), U}) = convert(T, X)
```

Comment:
If I break this up a bit, I like being able to convert to / from:

NTuples - to cut down on some syntax, e.g.
```
immutable Foo{T}
     q::Quaternion{T}
    ...
```
I find:
```    Foo((1.0, 0.0, 0.0, 0.0), ....)   ```
"cleaner" than
```    Foo(Quaternion(1.0, 0.0, 0.0, 0.0), ....)  ```
when Foo has lots of fields

Vec - to treat the parameters as just a vector in a space that you can
manipulate however you like (e.g. during optimization)

Vector - allowing a conversion a common interface for compatibility with
other packages

Putting this into the parent packages would be nice but there's no time
frame on the owner making another tagged version of their package


On 10 May 2016 at 15:05, Chris Foster <notifications@github.com> wrote:

> In src/type_methods.jl
> <https://github.com/FugroRoames/Rotations.jl/pull/2#discussion_r62612907>:
>
> > +        # convert to an immutable vector
> > +        @inline convert(::Type{Vec}, X::$(rot_type)) = Vec($(untyped_field_expr.args...))
> > +        @inline convert{T <: Real}(::Type{Vec{$(numel(rot_type)), T}}, X::$(rot_type)) = Vec($(typed_field_expr.args...))
> > +        @inline call(::Type{Vec}, X::$(rot_type)) = convert(Vec, X)
> > +        @inline call{T <: Real}(::Type{Vec{$(numel(rot_type)), T}}, X::$(rot_type)) = convert(Vec{$(numel(rot_type)), T}, X)
> > +
> > +        # convert from an immutable vector
> > +        @inline convert{T <: $(rot_type), U <: Real}(::Type{T}, X::FixedVector{$(numel(rot_type)), U}) = T($(untyped_indexed_expr.args...))
> > +        @inline call{T <: $(rot_type), U <: Real}(::Type{T}, X::FixedVector{$(numel(rot_type)), U}) = convert(T, X)
> > +
> > +        # convert to a tuple
> > +        # @inline tuple(X::$(rot_type)) = tuple($(untyped_field_expr.args...))
> > +
> > +        # convert from a tuple
> > +        @inline convert{T <: $(rot_type), U <: Real}(::Type{T}, X::NTuple{$(numel(rot_type)), U}) = T($(untyped_indexed_expr.args...))
> > +        @inline call{T <: $(rot_type), U <: Real}(::Type{T}, X::NTuple{$(numel(rot_type)), U}) = convert(T, X)
>
> It seems a somewhat questionable to be adding so many generic constructors
> and conversions to types which aren't defined in this module, eg
> Quaternion and RotMatrix at least.
>
> Is the idea that you want to flatten the storage so you can stick it into
> a larger array for optimization? It might be better to cut the methods back
> here, perhaps implement vec to flatten the types? (vec() for FSAs should
> go into FixedSizeArrays, presumably returning an appropriate Vec from any
> other FSA).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/FugroRoames/Rotations.jl/pull/2/files/fb3a7882257b93afed97cfaa9bf56669bbf410dc#r62612907>
>